sw/device/lib/runtime/pmp.c

```diff
@@ -0,0 +1,539 @@
+    case kPmpRegionPermissionsReadWriteExecute:
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_R, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_W, true);
+      *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_X, true);
+      break;
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+// TODO - short description.
+static void pmp_cfg_mode_lock_set(pmp_region_lock_t lock, uint32_t *bitfield) {
+  if (lock == kPmpRegionLockLocked) {
+    *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_L, true);
+  } else {
+    *bitfield = bitfield_bit32_write(*bitfield, PMP_CFG_CSR_L, false);
+  }
+}
+
+/**
+ * Check whether `address` is correctly aligned.
+ *
+ * The alignment depend on the granularity, which is implementation specific,
+ * and for Ibex is `PMP_GRANULARITY_IBEX`. Default granularity "G" is 0, which
+ * means a minimal alignment of 4bytes. Please see:
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6 Physical Memory Protection", "Figure 3.26" and section
+ * "Address Matching".
+ *
+ * @param address Conventional system address.
+ * @return `true` on success, `false` on failure.
+ */
+static bool pmp_address_aligned(pmp_region_address_t address) {
+  return address == (address & PMP_ADDRESS_ALIGNMENT_INVERTED_MASK);
+}
+
+/**
+ * Constructs a NAPOT address from the requested conventional address and size.
+ *
+ * This function makes sure that the `address` and `size` are valid, and then
+ * constructs a corresponding NAPOT address. Please see:
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6 Physical Memory Protection", "Figure 3.26" and "Table 3.10".
+ *
+ * @param address Conventional system address.
+ * @param size The size of a range to protect.
+ * @param napot_address Constructed NAPOT address.
+ * @return `true` on success, `false` on failure.
+ */
+static bool pmp_napot_address_construct(pmp_region_address_t address,
+                                        uint32_t size,
+                                        pmp_region_address_t *napot_address) {
+  // Must be at least the size of the minimal alignment adjusted for
+  // granularity, and the minimal allowed size for the NAPOT mode.
+  if (size < PMP_ADDRESS_ALIGNMENT || size < PMP_ADDRESS_MIN_ALIGNMENT_NAPOT) {
+    return kPmpRegionConfigureNapotBadAddress;
+  }
+
+  // Check if the `size` if Power Of Two.
+  uint32_t size_mask = size - 1;
+  if ((size & size_mask) != 0) {
+    return kPmpRegionConfigureNapotBadSize;
+  }
+
+  // Check if the address is aligned to the `size`.
+  uint32_t size_mask_inverted = ~size_mask;
+  if (address == (address & size_mask_inverted)) {
+    return kPmpRegionConfigureNapotBadAddress;
+  }
+
+  // `size` is Power Of Two, so the invert can be used to clear the size bit of
+  // the NAPOT address.
+  uint32_t size_inverted = ~size;
+
+  *napot_address = address | size_inverted;
+  *napot_address &= size_mask;
```

Comment:
This line seems unnecessary, since we already know that `address` is `size`-aligned?

Reply:
This, sets the `size - 1` bits that otherwise are `0`.

Reply:
Now, this one is supposed to be `|`.

Reply:
Done.